### PR TITLE
[FIX][I] Safe document contents before negotiation

### PR DESCRIPTION
**Pull request by tobous**
To avoid inconsistencies, all open documents are saved on the client
side before starting the project negotiation.

This is more of a temporary solution as this should ideally be handled
by the core logic as well. Such a logic was not needed until now as
Saros/E subverts this issue by closing all open editors on the client
side before the project negotiation starts, thereby saving all modified
files.

### CODE REVIEW

File: de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/ui/wizards/AddProjectToSessionWizard.java [hunk 1]
```diff
@@ -104,6 +105,9 @@
     @Inject
     private IWorkspace workspace;
 
+    @Inject
+    private ProjectAPI projectAPI;
```
Comment (saros-infrastructure) [inline de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/ui/wizards/AddProjectToSessionWizard.java]: Code Smell Code Smell: Make "projectAPI" transient or serializable. ([squid:S1948](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS1948))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=206&issues=AWUqRKwvW2AOP0Ck732W&open=AWUqRKwvW2AOP0Ck732W)

[ISSUE_KEY]: AWUqRKwvW2AOP0Ck732W

Comment (saros-infrastructure) [inline de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/ui/wizards/AddProjectToSessionWizard.java]: Code Smell Code Smell: Make "projectAPI" transient or serializable. ([squid:S1948](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS1948))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=206&issues=AWUqRKwvW2AOP0Ck732W&open=AWUqRKwvW2AOP0Ck732W)

[ISSUE_KEY]: AWUqRKwvW2AOP0Ck732W

File: de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/core/ui/util/CollaborationUtils.java [hunk 2]
```diff
@@ -74,6 +78,8 @@ private CollaborationUtils() {
     public static void startSession(List<IResource> resources,
         final List<JID> contacts) {
 
+        projectAPI.saveAllDocuments();
```
Comment (srossbach) [inline de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/core/ui/util/CollaborationUtils.java]: This should not be necessary.

Comment (tobous) [reply de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/core/ui/util/CollaborationUtils.java]: You are right, this is not needed. My bad. Must have misread my test results.

**Comment by tobous** [comment]
The editors on the receiving side are the problem. I don't think closing the editors on the receiving side is a good solution for the problem. As long as we don't have a solution in the core for this, I would like to add this fix (the part in `AddProjectToSessionWizard`).

But I still think this belongs into the core. I would go ahead and create a ticket for this if you don't disagree.
Saving all open editors might be trivial during the IncomingProjectNegotiation, as the needed methods already exist. But I don't know if we have the needed resources during at that time. The save function used in [OutgoingProjectNegotiation](https://github.com/saros-project/saros/blob/master/de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/negotiation/AbstractOutgoingProjectNegotiation.java#L338) needs an IProject object for the shared project/module.

Regarding the VFS reload:
Do you know how the unpacking of the transferred archive works during the project negotiation? Are the resources created using `IFile.create(...)`/`IFolder.create(...)`? This would make the second commit of this PR unnecessary as well. The create method uses the IntelliJ VFS to create the files, meaning the VFS does not have to be refreshed after the negotiation.

**Comment by srossbach** [comment]
Writing is done here https://github.com/saros-project/saros/blob/master/de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/negotiation/DecompressArchiveTask.java#L114

Although I am not quite sure if the IPN already tries to create the necessary folder before.

I think there are already some TODOs regarding write transactions (I think this is only possible in Eclipse without freezing the UI) and of course the dirty hack to the FileReplacementInProgress observeable (which can produce other defects as it stalls all resource tracking, at least for Eclipse).

File: de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/ui/wizards/AddProjectToSessionWizard.java (hunk 1 above)
Comment (saros-infrastructure) [inline de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/ui/wizards/AddProjectToSessionWizard.java]: Code Smell Code Smell: Make "projectAPI" transient or serializable. ([squid:S1948](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS1948))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=206&issues=AWUqRKwvW2AOP0Ck732W&open=AWUqRKwvW2AOP0Ck732W)

[ISSUE_KEY]: AWUqRKwvW2AOP0Ck732W

Comment (saros-infrastructure) [inline de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/ui/wizards/AddProjectToSessionWizard.java]: Code Smell Code Smell: Make "projectAPI" transient or serializable. ([squid:S1948](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS1948))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=206&issues=AWUqRKwvW2AOP0Ck732W&open=AWUqRKwvW2AOP0Ck732W)

[ISSUE_KEY]: AWUqRKwvW2AOP0Ck732W

**Pull request merged**